Updates vyper versions to supported versions (#586)
matter-labs/zksync-web-era-docs: docs/tools/hardhat/hardhat-zksync-vyper.md

# `hardhat-zksync-vyper`

The [@matterlabs/hardhat-zksync-vyper](https://www.npmjs.com/package/@matterlabs/hardhat-zksync-vyper) plugin provides an interface for compiling Vyper smart contracts before deploying them to zkSync Era.

## Set up

### 1. Scaffold a new project

Use the [zkSync Era cli](../../tools/zksync-cli/README.md) to set up a project.

```sh

> npx zksync-cli@latest create greeter-vyper-example

Reviewer: ```suggestion
npx zksync-cli@latest create greeter-vyper-example
cd greeter-vyper-example
```

    },
    zkSyncTestnet: {
      url: "https://testnet.era.zksync.dev",
      ethNetwork: "goerli",
      zksync: true,
    },
  },
  // Currently, only Vyper ^0.3.3 is supported.
  vyper: {
    version: "0.3.3",
  },
};

export default config;
```

#### Options

- `version`: The `zkvyper` compiler version. Find the latest compiler versions in the [zkvyper repo](https://github.com/matter-labs/zkvyper-bin).
- `compilerSource`: Indicates the compiler source and can be either `binary`. (A `docker` option is no longer recommended). If there is no previous installation, the plugin automatically downloads one. 

:::warning
The `docker` option is not recommended as compilers are no longer released as Docker images.
:::

- `compilerPath`: Optional field with the path to the `zkvyper` binary. By default, the binary in `$PATH` is used.
- `libraries`: Define any non-inlinable libraries your contracts use as dependencies here. Learn more about [compiling libraries](./compiling-libraries.md).
- `zksync`: Indicates whether `zkvyper` is enabled on zkSync Era. This option is useful for multichain projects in which you want to enable `zksync` for specific networks only.

> ### 4. Set Up Vyper contracts

Reviewer: ```suggestion
### 4. Create Vyper contract
```

The [zkSync Era cli](../../tools/zksync-cli/README.md) generates a `contracts` folder which includes a `Greeter.sol` contract. 

- Delete `Greeter.sol` from the `contracts/` directory.
- Add the equivalent `Greeter.vy` Vyper contract:

```vyper
# @version ^0.3.3
# vim: ft=python

owner: public(address)
greeting: public(String[100])

# __init__ is not invoked when deployed from create_forwarder_to
@external
def __init__(greeting: String[64]):
  self.owner = msg.sender
  self.greeting = greeting

# Invoke once after create_forwarder_to
@external
def setup(_greeting: String[100]):
  assert self.owner == ZERO_ADDRESS, "owner != zero address"
  self.owner = msg.sender
  self.greeting = _greeting

@external
@view
def greet() -> String[100]:
    return self.greeting
```

### 3. Compile the contract 

::: code-tabs
@tab:active yarn
```bash
yarn hardhat compile
```
@tab npm
```bash
npx hardhat compile
```
:::

> ### 4. Deploy the contract

Reviewer: ```suggestion
### 6. Deploy the contract
```

First update the `use-greeter.ts` script, supplied by the CLI in the `deploy/` directory. 

Alter this line:
```
// Load contract artifact. Ensure to compile first!
import * as ContractArtifact from "../artifacts-zk/contracts/Greeter.sol/Greeter.json";
```

To aim at our Vyper contract:
```
// Load contract artifact. Ensure to compile first!
import * as ContractArtifact from "../artifacts-zk/contracts/Greeter.vy/Greeter.json";
```

> ### 5. Add private key to environment variables

Reviewer: ```suggestion
### 7. Add private key to environment variables
```

Remove `example` from the `.env.example` file and add your private key to `<WALLET-PRIVATE-KEY>`.

> 6. Deploy the contract

Reviewer: ```suggestion
8. Deploy the contract
```

  self.owner = msg.sender
  self.greeting = greeting

# Invoke once after create_forwarder_to
@external
def setup(_greeting: String[100]):
  assert self.owner == ZERO_ADDRESS, "owner != zero address"
  self.owner = msg.sender
  self.greeting = _greeting

@external
@view
def greet() -> String[100]:
    return self.greeting
```

### 5. Compile the contract 

::: code-tabs
@tab:active yarn
```bash
yarn hardhat compile
```
@tab npm
```bash
npx hardhat compile
```
:::

### 6. Create deployment script

First update the `use-greeter.ts` script, supplied by the CLI in the `deploy/` directory. 

Alter this line:
```
// Load contract artifact. Ensure to compile first!
import * as ContractArtifact from "../artifacts-zk/contracts/Greeter.sol/Greeter.json";
```

To aim at our Vyper contract:
```
// Load contract artifact. Ensure to compile first!
import * as ContractArtifact from "../artifacts-zk/contracts/Greeter.vy/Greeter.json";
```

### 7. Add private key to environment variables

Remove `example` from the `.env.example` file and add your private key to `<WALLET-PRIVATE-KEY>`.

8. Deploy the contract

```
yarn hardhat deploy-zksync --script deploy-greeter.ts
```

> ### 9. Output

Reviewer: ```suggestion
### 7. Output
```

> ### 7. Output

Reviewer: ```suggestion
### 9. Output
```

- `libraries`: Define any non-inlinable libraries your contracts use as dependencies here. Learn more about [compiling libraries](./compiling-libraries.md).
- `zksync`: Indicates whether `zkvyper` is enabled on zkSync Era. This option is useful for multichain projects in which you want to enable `zksync` for specific networks only.

> ### 4. Set Up Vyper Contracts

Reviewer: ```suggestion
### 4. Set Up Vyper contracts
```

> To prepare Vyper contracts, leverage the [zkSync Era cli](../../tools/zksync-cli/README.md), which initially generates a `Contracts` folder including a Greeter.Sol contract. As our objective involves compiling Vyper contracts, not Solidity ones, modifications are in order. 

Reviewer: ```suggestion
The [zkSync Era cli](../../tools/zksync-cli/README.md) generates a `contracts` folder which includes a `Greeter.sol` contract. 
```

> - Delete Greeter.Sol from the `Contracts/` directory.

Reviewer: ```suggestion
- Delete `Greeter.sol` from the `contracts/` directory.
```

- Add the equivalent `Greeter.vy` Vyper contract:

```vyper
# @version ^0.3.3
# vim: ft=python

owner: public(address)
greeting: public(String[100])

# __init__ is not invoked when deployed from create_forwarder_to
@external
def __init__(greeting: String[64]):
  self.owner = msg.sender
  self.greeting = greeting

# Invoke once after create_forwarder_to
@external
def setup(_greeting: String[100]):
  assert self.owner == ZERO_ADDRESS, "owner != zero address"
  self.owner = msg.sender
  self.greeting = _greeting

@external
@view
def greet() -> String[100]:
    return self.greeting
```

> ### 3. Compile Your Contract 

Reviewer: ```suggestion
### 3. Compile the contract 
```

> You can compile the contract with the following command:

Reviewer: ```suggestion
```

@tab npm
```bash
npx hardhat compile
```
:::

> ### 4. Deploy Your Contract

Reviewer: ```suggestion
### 4. Deploy the contract
```

> To deploy the `Greeter.vy` contract, initially update the `use-greeter.ts` script, supplied by the CLI in the `deploy/` directory. 

Reviewer: ```suggestion
First update the `use-greeter.ts` script, supplied by the CLI in the `deploy/` directory. 
```

Alter this line:
```
// Load contract artifact. Ensure to compile first!

> import * as ContractArtifact from "../artifacts-zk/contracts/Greeter.Sol/Greeter.json";

Reviewer: ```suggestion
import * as ContractArtifact from "../artifacts-zk/contracts/Greeter.sol/Greeter.json";
```

Reviewer: is it really a capital S?

Author: Nope my mistake.

Reviewer: i only asked because it was so consistent ... :))))

```

To aim at our Vyper contract:
```
// Load contract artifact. Ensure to compile first!
import * as ContractArtifact from "../artifacts-zk/contracts/Greeter.vy/Greeter.json";
```

> Following that, substitute `<WALLET-PRIVATE-KEY>` in the `deploy/deploy-greeter.ts` with your personal key. Now, you're ready to deploy the `Greeter.vy` contract! 

Reviewer: ```suggestion
### 5. Add private key to environment variables

Remove `example` from the `.env.example` file and add your private key to `<WALLET-PRIVATE-KEY>`.
```

> ```

Reviewer: ```suggestion
6. Deploy the contract

```
```

> yarn hardhat deploy-zksync

Reviewer: OK, when I run this I get errors.

```
% yarn hardhat deploy-zksync
yarn run v1.22.19
warning ../../../../package.json: No license field
$ /Users/niramisaweiss/Documents/plugins/vyper-npm/greeter-vyper-example/node_modules/.bin/hardhat deploy-zksync
Error HH303: Unrecognized task deploy-zksync

For more info go to https://hardhat.org/HH303 or run Hardhat with --show-stack-traces
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
```

Reviewer: OK, checked, made some updates, the numbering went a bit haywire..

```

> ### 5. Expect the Following Output

Reviewer: ```suggestion
### 6. Output
```

> You should anticipate the following output:

Reviewer: ```suggestion
You should see something like this:
```
